File: mobile/src/api/pagination.ts
import type { PaginatedResponse, PaginationParams } from '@jot/shared';

const MOBILE_PAGE_SIZE = 100;

export async function collectAllPages<T>(
  fetchPage: (params: PaginationParams) => Promise<PaginatedResponse<T>>,
): Promise<T[]> {
  const items: T[] = [];
  let offset = 0;

  for (;;) {
    const response = await fetchPage({ limit: MOBILE_PAGE_SIZE, offset });
    items.push(...response.items);

    if (!response.pagination.has_more) {
      return items;
    }

    if (response.pagination.next_offset === undefined) {
      throw new Error('Paginated response is missing next_offset');
    }

    offset = response.pagination.next_offset;
  }
}

File: server/client/pagination.go
package client

import (
	"net/url"
	"strconv"
)

const maxPageSize = 100

func applyPagination(q url.Values, opts *PaginationOptions) {
	if opts == nil {
		return
	}
	if opts.Limit > 0 {
		q.Set("limit", strconv.Itoa(opts.Limit))
	}
	if opts.Offset > 0 {
		q.Set("offset", strconv.Itoa(opts.Offset))
	}
}

func collectAllPages[T any](fetchPage func(*PaginationOptions) (*PaginatedResponse[T], error)) ([]T, error) {
	opts := &PaginationOptions{Limit: maxPageSize}
	items := make([]T, 0)

	for {
		page, err := fetchPage(opts)
		if err != nil {
			return nil, err
		}

		items = append(items, page.Items...)
		if !page.Pagination.HasMore || page.Pagination.NextOffset == nil {
			return items, nil
		}

		opts.Offset = *page.Pagination.NextOffset
	}
}
